README.md
# smooth-keyframes

Generate a curve that smoothly interpolates between a set of keyframes, using a monotonicity-preserving cubic spline as proposed in [A simple method for monotonic interpolation in one dimension](http://adsabs.harvard.edu/full/1990A%26A...239..443S) and implemented in [d3-shape](https://github.com/d3/d3-shape/blob/master/src/curve/monotone.js).

Unlike normal splines, this smoothly connects a series of points without *spurious oscillations*. You can see what that means here: https://smooth-keyframes-demo.surge.sh


## API

```js
import keyframes from 'smooth-keyframes';

const y = keyframes([
	[0,0],
	[50, 100],
	[100, 50]
]);

// get the y at a given x
y(-1); // 0 — input values are clamped
y(0); // 0
y(25); // 50
y(30); // 64.8
y(50); // 100
y(100); // 50
y(101); // 50 — input values are clamped
```


## License

[MIT](LICENSE)


CHANGELOG.md
# spline-keyframes changelog

## 1.0.0

* Rewrite, update API

## 0.0.5

* Fix broken binary search that would occasionally result in infinite loop

## 0.0.4

* Binary search
* Concentrate sampling on curviest segments, omit redundant samples

## 0.0.3

* Fix NaN bug with consective frames with no change in value

## 0.0.2

* Use a different set of constraints to avoid 'overshoot' and bad curves
* Changed API. Curves are now immutable.

## 0.0.1

* First release

test/test.ts
import * as assert from 'assert';
import keyframes from '../src/index';

require('console-group').install();

describe('smooth-keyframes', () => {
	it('errors if no frames provided', () => {
		assert.throws(() => {
			keyframes([]);
		}, /You must provide at least one point/);
	});

	it('handles a single point', () => {
		const y = keyframes([[5, 5]]);

		assert.equal(y(0), 5);
		assert.equal(y(5), 5);
		assert.equal(y(10), 5);
	});

	it('handles two points', () => {
		const y = keyframes([[4, 4], [6, 6]]);

		assert.equal(y(4), 4);
		assert.equal(y(5), 5);
		assert.equal(y(6), 6);
	});

	it('returns an intermediate value', () => {
		const y = keyframes([
			[0, 0],
			[1, 100],
			[2, 50]
		]);

		const v = y(0.5);
		assert.ok(0 < v && v < 100);
	});

	it('clamps values', () => {
		const y = keyframes([
			[0, 10],
			[1, 20],
			[2, 30]
		]);

		assert.equal(y(-1), 10);
		assert.equal(y(3), 30);
	});
});

src/index.ts
function sign(x: number) {
	return x < 0 ? -1 : 1;
}

// Adapted from https://github.com/d3/d3-shape/blob/master/src/curve/monotone.js
// https://github.com/d3/d3-shape/blob/master/LICENSE

// Calculate the slopes of the tangents (Hermite-type interpolation) based on
// the following paper: Steffen, M. 1990. A Simple Method for Monotonic
// Interpolation in One Dimension. Astronomy and Astrophysics, Vol. 239, NO.
// NOV(II), P. 443, 1990.
function slope3(x0: number, x1: number, x2: number, y0: number, y1: number, y2: number) {
	var h0 = x1 - x0,
		h1 = x2 - x1,
		s0 = (y1 - y0) / (h0 || (h1 < 0 && -0)),
		s1 = (y2 - y1) / (h1 || (h0 < 0 && -0)),
		p = (s0 * h1 + s1 * h0) / (h0 + h1);
	return (
		(sign(s0) + sign(s1)) *
			Math.min(Math.abs(s0), Math.abs(s1), 0.5 * Math.abs(p)) || 0
	);
}

export default function keyframes(frames: Array<[number, number]>) {
	if (!frames || frames.length === 0) {
		throw new Error(`You must provide at least one point`);
	}

	frames = frames.slice().sort((a, b) => a[0] - b[0]);

	const segments: Array<{
		x0: number,
		y0: number,
		x1: number,
		y1: number,
		width: number,
		gradient: number,
		tangent0: number,
		tangent1: number
	}> = [];

	let x0 = NaN;
	let x1 = NaN;
	let y0 = NaN;
	let y1 = NaN;
	let tangent0 = NaN;
	let state = 1;

	function add_segment(tangent0: number, tangent1: number) {
		segments.push({
			x0,
			y0,
			x1,
			y1,
			width: (x1 - x0),
			gradient: (y1 - y0) / (x1 - x0),
			tangent0,
			tangent1
		});
	}

	frames.forEach(point => {
		const [x2, y2] = point;

		let tangent1 = NaN;

		switch (state) {
			case 1:
				state = 2;
				break;
			case 2:
				state = 3;
				tangent1 = slope3(x0, x1, x2, y0, y1, y2);
				break;
			default:
				tangent1 = slope3(x0, x1, x2, y0, y1, y2)
				add_segment(tangent0, tangent1);
				break;
		}

		x0 = x1, x1 = x2;
		y0 = y1, y1 = y2;
		tangent0 = tangent1;
	});

	const first = frames[0];
	const last = frames[frames.length - 1];

	add_segment(tangent0, 0); // final tangent, like starting tangent, is 0

	return (x: number) => {
		if (x <= first[0]) return first[1];
		if (x >= last[0]) return last[1];

		let low = 0;
		let high = segments.length - 1;

		while (low <= high) {
			const mid = (low + high) >> 1;
			const segment = segments[mid];

			if (segment.x0 <= x && x < segment.x1) {
				const dx0 = (x - segment.x0);
				const dx1 = (x - segment.x1);

				const t = dx0 / segment.width;

				let y = segment.y0 + dx0 * segment.gradient;

				const influence0 = (segment.y0 + dx0 * segment.tangent0) - y;
				const influence1 = (segment.y1 + dx1 * segment.tangent1) - y;

				return y + (
					((1 - t) ** 2) * influence0 +
					(t ** 2) * influence1
				);
			}

			if (x < segment.x0) {
				high = mid - 1;
			} else {
				low = mid + 1;
			}
		}
	};
}


src/Segment.ts
import { Keyframe } from './Keyframe';
import { Vector } from './Vector';

export class Segment {
	start: number;
	end: number;
	duration: number;

	p0: Vector;
	p1: Vector;
	p2: Vector;
	p3: Vector;

	p0_p1: Vector;
	p1_p2: Vector;
	p2_p3: Vector;

	magnitude: number;

	constructor(p0: Keyframe, p3: Keyframe) {
		this.start = p0.time;
		this.end = p3.time;
		this.duration = this.end - this.start;

		this.p0 = new Vector(p0.time, p0.value);
		this.p3 = new Vector(p3.time, p3.value);
		this.p1 = this.p0.add(p0.control[1]);
		this.p2 = this.p3.add(p3.control[0]);

		this.magnitude = this.p3.subtract(this.p0).magnitude();
	}

	addPoints(n: number, points: Array<{ time: number, value: number }>) {
		const { p0, p1, p2, p3 } = this;

		if (p3.value === p0.value) n = 1;

		for (let t = 0; t < 1; t += 1 / n) {
			// https://stackoverflow.com/a/8218244/2742396
			const time = (
				(p0.time * Math.pow(1 - t, 3)) +
				(p1.time * 3 * Math.pow(1 - t, 2) * t) +
				(p2.time * 3 * (1 - t) * Math.pow(t, 2)) +
				(p3.time * Math.pow(t, 3))
			);

			const value = (
				(p0.value * Math.pow(1 - t, 3)) +
				(p1.value * 3 * Math.pow(1 - t, 2) * t) +
				(p2.value * 3 * (1 - t) * Math.pow(t, 2)) +
				(p3.value * Math.pow(t, 3))
			);

			points.push({ time, value });
		}
	}
}


src/Keyframe.ts
import { Curve } from './Curve';
import { Vector } from './Vector';

export type KeyframeData = {
	time: number;
	value: number;
	tension?: number;
}

function clamp(n: number, min: number, max: number) {
	return n < min ? min : n > max ? max : n;
}

export class Keyframe {
	curve: Curve;
	time: number;
	value: number;
	tension: number;

	previous: Keyframe;
	next: Keyframe;

	control: [Vector, Vector];

	constructor(curve: Curve, data: KeyframeData) {
		this.curve = curve;
		this.time = data.time;
		this.value = data.value;
		this.tension = curve.tension * ('tension' in data ? data.tension : 1);
	}

	_link(previous: Keyframe, next: Keyframe) {
		this.previous = previous;
		this.next = next;

		this.control = [null, null];

		if (previous && next) {
			const line = new Vector(next.time - previous.time, next.value - previous.value).scale(this.tension);

			const d0 = new Vector(this.time - previous.time, this.value - previous.value).magnitude();
			const d1 = new Vector(next.time - this.time, next.value - this.value).magnitude();

			this.control[0] = line.scale(-d0 / (d0 + d1));
			this.control[1] = line.scale(d1 / (d1 + d0));

			const value_multiplier = clamp(Math.min(
				0.5 * (previous.value - this.value) / this.control[0].value,
				0.5 * (next.value - this.value) / this.control[1].value
			) || 0, 0, 1);

			const time_multiplier = Math.min(
				0.5 * (previous.time - this.time) / this.control[0].time,
				0.5 * (next.time - this.time) / this.control[1].time,
				1
			);

			this.control[0].time *= time_multiplier;
			this.control[1].time *= time_multiplier;

			this.control[0].value *= value_multiplier;
			this.control[1].value *= value_multiplier;
		} else if (previous) {
			this.control[0] = new Vector(previous.time - this.time, 0).scale(Math.min(this.tension, 0.5));
		} else if (next) {
			this.control[1] = new Vector(next.time - this.time, 0).scale(Math.min(this.tension, 0.5));
		}
	}
}

src/Curve.ts
import { Keyframe, KeyframeData } from './Keyframe';
import { Segment } from './Segment';

export type CurveOptions = {
	tension?: number;
	samples?: number;
}

export class Curve {
	tension: number;
	samples: number;
	frames: Keyframe[];
	segments: Segment[];
	points: Array<{ time: number, value: number }>;

	constructor(frames: KeyframeData[], opts?: CurveOptions) {
		this.tension = opts && 'tension' in opts ? opts.tension : 0.5;
		this.samples = opts && 'samples' in opts ? opts.samples : 100;

		this.frames = frames
			.sort((a: Keyframe, b: Keyframe) => {
				if (a.time === b.time) {
					throw new Error('Cannot have multiple keyframes with the same `time`');
				}

				return a.time - b.time;
			})
			.filter((frame, i) => {
				const previous = frames[i - 1];
				const next = frames[i + 1];
				if (!previous || !next) return true; // always include endpoints
				return (previous.value !== frame.value) || (next.value !== frame.value);
			})
			.map((data: KeyframeData) => new Keyframe(this, data));

		this.frames.forEach((frame, i) => {
			frame._link(this.frames[i - 1], this.frames[i + 1]);
		});

		this.segments = [];
		for (let i = 0; i < this.frames.length - 1; i += 1) {
			this.segments[i] = new Segment(
				this.frames[i],
				this.frames[i + 1]
			);
		}

		const first = this.frames[0];
		const last = this.frames[this.frames.length - 1];

		const start = first.time;
		const end = last.time;
		const duration = end - start;

		const n = opts && 'samples' in opts ? opts.samples : duration / (1 / 30);
		const step = (end - start) / n;

		this.points = [];

		const total_magnitude = this.segments.reduce((m, segment) => m + segment.magnitude, 0);

		this.segments.forEach(segment => {
			segment.addPoints(
				Math.floor(n * segment.magnitude / total_magnitude),
				this.points
			);
		});

		this.points.push({
			time: last.time,
			value: last.value
		});
	}

	at(time: number) {
		if (typeof time !== 'number' || isNaN(time)) throw new Error('Expected a number');

		const first = this.frames[0];
		const last = this.frames[this.frames.length - 1];

		if (time <= first.time) return first.value;
		if (time >= last.time) return last.value;

		let low = 0;
		let high = this.points.length - 2;

		while (low <= high) {
			const mid = (low + high) >> 1;
			const a = this.points[mid];
			const b = this.points[mid + 1];

			if (a.time <= time && time < b.time) {
				const p = (time - a.time) / (b.time - a.time);
				return a.value + p * (b.value - a.value);
			}

			if (time < a.time) {
				high = mid - 1;
			} else {
				low = mid + 1;
			}
		}
	}
}